### PR TITLE
Patterns: Improve the screenreader announcements for category navigation

### LINKED ISSUE #52005: Library: Improve the screenreader announcements for category navigation
### Description

When tabbing through the category navigation on the Library screen, the way categories are announced could be improved:
Screen Shot 2023-06-28 at 11 31 49 am

Currently these will be announced as "Headers 2, button", "Footers 1, button", ... "Custom Patterns 4, current item, button".

It's good that the selected item is announced, but a couple of things could be improved:
- There are icons that relay that a category is for a template part or a pattern visually (though it's very subtle), but this information isn't relayed textually.
- The counts of items in the category is useful, but might be confusing because of the way it's read out.

It might be a good idea to look at adding some `VisuallyHidden` text to improve the announcements to be something like:

"Headers (template part), 2 items, button", ... "Custom patterns (pattern), 4 items, current item, button"

### Step-by-step reproduction instructions

1. Open the site editor
2. Tab to the Library option
3. Tab through the categories to hear the announcements.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by carolinan**
## What?

Updated 11th of July 2023.
Closes https://github.com/WordPress/gutenberg/issues/52005

When the original issue was opened, the Site Editor patterns page did not include information about if a category was for a pattern or a template part. The original PR aimed to address that.
Now, the patterns page has visible h2 headings above the different categories. These headings are not focusable but are announced by screen readers when reading the full content of the sidebar (navigation mode).

The updated PR attempts to implement the suggestion from the comment below:

1. hide the count with aria-hidden="true"
2. add a VisuallyHidden component outside of the button that holds a descriptive text e.g.: template part, 1 item
3. reference the visually hidden description by the means of aria-describedby
4. this would require using a unique ID, probably by the means of useInstanceId.

### Testing Instructions for Screen readers
Make sure that your WordPress install has some synced block patterns.
Activate a block theme, and the screen reader.
Open the Patterns page via the WordPress admin area > Appearance > Editor > Patterns.
Use the tab key to navigate past the initial links and buttons until you have passed the "Create pattern" button.

Using VoiceOver, navigating with tab key: Announces "My patterns, button.... more content available",  Opening the more content menu announces  "1" - the number of items in the category.
Pressing the tab key again will skip past the non-focusable heading for the template parts, and announce:
"Header, button .... more content available". Opening the more content menu announces  "1" - the number of items in the category.

Using Voiceover, reading the full content of the sidebar:
Announces  "My patterns, My patterns, button, 1" followed by "Heading level 2, Template parts, list, 3 items, Header, current item, Header, button, 2".
So in this case, there is a lot of repetition.

### CODE REVIEW

**Comment by aristath** [comment]
cc @afercia

File: packages/edit-site/src/components/sidebar-navigation-screen-library/category-item.js [hunk 1]
```diff
@@ -30,12 +36,21 @@ export default function CategoryItem( {
 		return;
 	}
 
+	const ariaLabel = sprintf(
+		/* translators: %1$s is the category name, %2$s is the type (pattern or template part), %3$s is the number of items. */
+		_n( '%1$s (%2$s), %3$s item', '%1$s (%2$s), %3$s items', count ),
```
Comment (afercia) [inline packages/edit-site/src/components/sidebar-navigation-screen-library/category-item.js]: This string would be a bit complicated to translate. It contains a variable part (the type) that may vary in gender on same languages. Ideally, we should try to follow bets practices here and try to provide separate strings with only the count as variable part.

File: packages/edit-site/src/components/sidebar-navigation-screen-library/category-item.js [hunk 2]
```diff
@@ -30,12 +36,21 @@ export default function CategoryItem( {
...
+		count
+	);
+
 	return (
 		<SidebarNavigationItem
 			{ ...linkInfo }
 			icon={ icon }
 			suffix={ <span>{ count }</span> }
 			aria-current={ isActive ? 'true' : undefined }
+			aria-label={ ariaLabel }
```
Comment (afercia) [inline packages/edit-site/src/components/sidebar-navigation-screen-library/category-item.js]: This aria-label would make the accessible name pretty different from the visible text. While this adds contextual information for screen reader users, it would not help much speech recognition software users. For example:
VIsible text: `Footers 1`
Accessible name: `Footers (template part), 1 item`
Will detaile a bit better in a comment.

**Comment by afercia** [comment]
Besides the issue with the translatable string (See inline comment) I'm a bit worried that altering the accessible name by the means of an aria-label will introduce a barrier for speech recognition software users. The visible text and the actual accessible name would be pretty different. This would introduce a problem to issue appropriate voice commands. This problem is in part already present in the current implementation.

Testing with macOS Safari + Voice Control:

Before this PR:
visible text: Footers 1 (notice the number is placed on the right, far from the item name)

Screenshot 2023-06-29 at 14 55 52

accessible name: Footers 1

As a speech recognition software user, I'd first try a voice command like:
`Click Footers`
I'd do that because the number 1 is visually disconnected from the item name. Seems logical to just say `Footers`.
This voice command will fail because the actual accessible name is `Footers 1`.
However, after some trial and error, I'd likely try a voice command  like:
`Click Footers 1`.
This command works.
The viisble text matches the accessible name. So far so good.

Instead, after this patch:
The visible text is always the same: Footers 1
The actual accessible name is now: Footers (template part), 1 item

As a spech recognition software user, I'd try voice commands that match the visible text, as in the example above:
- `Click Footers`
- `Click Footers 1`

Both commands will fail because the actual accessible name is now very different. As a sighted user, I wouldn't have a clue what the accessible name is.

For this reason, we should always be very careful when altering the accessible name of a control.

In this case, I'd suggest to keep the accessible name unchanged and:
- hide the count with `aria-hidden="true"`
- add a `VisuallyHidden` component _outside_ of the button that holds a descriptive text e.g.: `template part, 1 item`
- reference the visually hidden description by the means of `aria-describedby`
- this would require using a unique ID, probably by the means of `useInstanceId`.

In this way, the items would provide more context for screen reader users anyways and the visible text would match the accessible name. Technically, an item would have:
- accessible name: `Footers`
- accessible description: `template part, 1 item`

Screen readers (with default settings) would read both, with the description being announced after a short delay:
`Footers` {short delay} `template part, 1 item`

**Comment by alexstine** [comment]
The only screen reader that would be effected by this is Voiceover. Voiceover exposes `aria-describedby` as a help tag where you must manually make Voiceover read the description. The UX is quite bad when trying to find out important information but it can be okay for information that is nice to know, but maybe not required. The real solution would be to have enough context in visible text where this isn't a problem but nothing in Gutenberg seems to ever be clear without hiding it behind a special accessibility attribute to keep the UI pretty.
1. @afercia already mentioned the problems with `aria-label`. Speech software is sadly not catching up fast enough and neither are the web specs.
2. The `aria-describedby` attribute can make it much harder for Voiceover users to get information readily available with Windows screen readers.

I think having the category and count as description text really does not fit the bill here and needs to be a main label somehow. Just not sure how we please everyone.

Sigh.

**Comment by carolinan** [comment]
I keep coming back to wanting to separate patterns and template parts in the main menu again. Even when knowing that separating them is temporary.  I don't disagree with any of the things said here as part of the reviews, I was merely trying to work with what we have.

**Comment by carolinan** [comment]
Maybe, the number of items is not actually needed, -I don't believe that empty categories are listed.

**Comment by alexstine** [comment]
@carolinan Sounds good. I think excluding number of items here is likely fine.

**Comment by bph** [comment]
@carolinan Are you still working on this for 6.3 or does this still need design input? 

cc @tellthemachines @ramonjd

**Comment by carolinan** [comment]
It still needs a new accessible design.

**Comment by carolinan** [comment]
As long as the needs design label is there, you can consider it as needing design.

**Review by aristath — APPROVED** [review]
Code looks good and this does improve things. 👍

**Comment by carolinan** [comment]
How can we bring attention to this accessibility bug so that it is fixed for the next minor WP release?
TLDR: The problem is that the current design, where the pattern category name and the number of items in that category are visually apart but in the same button element, makes it difficult to provide context to screen readers.
The fixes in this PR are not ideal.
